---
title: "Peak and motif analysis"
author: "`LMS Bioinformatics`"
date: "`r format(Sys.time(), '%B %Y')`"
output:
  slidy_presentation:
    highlight: tango
    keep_md: true
    footer: "MRC LMS Bioinformatics core"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)

# knitr::opts_chunk$set(cache = TRUE, warning = FALSE, echo=T,
#                       message = FALSE, cache.lazy = FALSE)

```


```{r, tidy=T, message=F, warning=F, echo=F}
setwd("/Volumes/bioinfomatics$/yfwang/BioinformaticsCore/GitHub/ATACseq_Practical_course")
```

```{r, tidy=T, message=F, warning=F}
library(rtracklayer)

peaks <- import("data/SRR3536937_1.trim.PE2SE.nodup.tn5_SRR3536938_1.trim.PE2SE.nodup.tn5.pf.narrowPeak.gz")
head(as.data.frame(peaks))

```

# header1

```{r, tidy=T, message=F, warning=F}

library(ChIPseeker)
library(org.Mm.eg.db)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)

peak_annotation <- annotatePeak(peaks, 
                                TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                                annoDb = "org.Mm.eg.db")

```

## plot1

```{r, tidy=T, message=F, warning=F}
plotAnnoBar(peak_annotation)
vennpie(peak_annotation)
upsetplot(peak_annotation)
upsetplot(peak_annotation, vennpie = T)
plotDistToTSS(peak_annotation)
hist(peak_annotation@anno$distanceToTSS, breaks = 2000, xlim = c(-50000, 50000))


```


# header2


```{r, tidy=T, message=F, warning=F}

# library(clusterProfiler)

# assignedGenes <- peak_annotation

```